script.sonos/resources/lib/speech.py

```diff
@@ -1,330 +1,327 @@
+            words = words.replace('%day', now.strftime('%A'))
+
+        if '%time' in words:
+            words = words.replace('%time', now.strftime('%H:%M'))
+
+        if '%date' in words:
+            d = int(now.strftime('%d'))
+            suffix = 'th' if 11 <= d <= 13 else {1: 'st', 2: 'nd', 3: 'rd'}.get(d % 10, 'th')
+            mth = now.strftime('%B')
+            words = words.replace('%date', str(d) + suffix + ' ' + mth)
+
+        if '%greet' in words:
+            GREET = [ADDON.getLocalizedString(32223),
+                     ADDON.getLocalizedString(32223),
+                     ADDON.getLocalizedString(32224),
+                     ADDON.getLocalizedString(32225)]
+            hour = int(now.strftime('%H'))
+            words = words.replace('%greet', GREET[hour / 6])
+
+        return words
+
+    # Show the keyboad so a user can type what speech string they want
+    def promptForInput(self):
+        header = "%s - %s" % (ADDON.getLocalizedString(32001), ADDON.getLocalizedString(32105))
+        keyboard = xbmc.Keyboard(heading=header)
+        keyboard.setHeading(header)
+        keyboard.doModal()
+
+        phrase = None
+        if keyboard.isConfirmed():
+            try:
+                phrase = keyboard.getText().decode("utf-8")
+            except:
+                phrase = keyboard.getText()
+            log("Speech: User input the text: %s" % phrase)
+
+        return phrase
+
+    # Loads all the saved phrases
+    def loadSavedPhrases(self):
+        # Get the location of the speech list file
+        configPath = xbmc.translatePath(ADDON.getAddonInfo('profile'))
+        speechfile = os_path_join(configPath, "speech.txt")
+        log("Speech: Phrases file location = %s" % speechfile)
+
+        phrases = []
+        # Check to see if the speech list file exists
+        if not xbmcvfs.exists(speechfile):
+            # Create a list of pre-defined phrases
+            phrases.append(ADDON.getLocalizedString(32221))
+            phrases.append(ADDON.getLocalizedString(32222))
+            phrases.append(ADDON.getLocalizedString(32226))
+            phrases.append('%greet')
+            phrases.sort()
+        else:
+            # Read the phases from the file
+            try:
+                file_in = open(speechfile, 'r')
+                phrases = file_in.readlines()
+                file_in.close()
+            except:
+                log("Speech: Failed to read lines from file %s" % speechfile, xbmc.LOGERROR)
+                log("Speech: %s" % traceback.format_exc(), xbmc.LOGERROR)
+
+        return phrases
+
+    # Writes the list of phrases to a file
+    def savePhrases(self, phrases):
+        # Get the location of the speech list file
+        configPath = xbmc.translatePath(ADDON.getAddonInfo('profile'))
+        speechfile = os_path_join(configPath, "speech.txt")
+        log("Speech: Phrases file location = %s" % speechfile)
+
+        try:
+            file_out = open(speechfile, 'w')

```

Comment:
please use the "with" statement to make sure that file gets closed properly again.